language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,230 @@
+        let mut removed_edges = BTreeMap::new();
+        for (n, es) in &self.0 {
+            removed_edges.insert(
+                n.clone(),
+                es.iter()
+                    .filter(|x| x.to == nonce)
+                    .cloned()
+                    .collect::<Vec<_>>(),
+            );
+        }
+        for (n, es) in &removed_edges {
+            es.iter().for_each(|removed_edge| {
+                let n_edge_set_ref = self.0.get_mut(n).unwrap();
+                n_edge_set_ref.remove(removed_edge);
+                nonce_edge_set.iter().for_each(|nonce_edge| {
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    } else {
+                        let edge = Edge {
+                            edge_type: EdgeType::Weak,
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
+                });
+            });
+        }
+        checked_postcondition!(!self.0.contains_key(&nonce));
+        checked_postcondition!(self.invariant());
+    }
+
+    /// renames nonces in `self` according to `nonce_map`
+    pub fn rename_nonces(&self, nonce_map: BTreeMap<Nonce, Nonce>) -> Self {
+        let mut new_graph = BTreeMap::new();
+        for (n, es) in &self.0 {
+            new_graph.insert(
+                nonce_map[n],
+                es.iter()
+                    .map(|x| Edge {
+                        edge_type: x.edge_type,
+                        label: x.label.clone(),
+                        to: nonce_map[&x.to],
+                    })
+                    .collect::<BTreeSet<_>>(),
+            );
+        }
+        let new_borrow_graph = BorrowGraph(new_graph);
+        checked_postcondition!(new_borrow_graph.invariant());
+        new_borrow_graph
+    }
+
+    /// checks if `self` is subsumed by `other`
+    pub fn subsumes(&self, other: &Self) -> bool {
+        self.unmatched_edges(other).values().all(|es| es.is_empty())
+    }
+
+    /// joins `other` into `self`
+    pub fn join(&mut self, other: &Self) {
+        for (n, es) in self.unmatched_edges(other) {
+            es.into_iter().for_each(|e| {
+                self.0.get_mut(&n).unwrap().insert(e);
+            })
+        }
+    }
+
+    /// gets all nonces that are targets of outgoing edges from `nonce`
+    pub fn all_borrows(&self, nonce: Nonce) -> BTreeSet<Nonce> {
```

Comment:
I found it helpful to have an API that returns the nil borrows, and all of the borrows rooted at the first label/field 
so something like
```
pub fn all_borrows(&self, id: RefID) -> (BTreeSet<RefID>, BTreeMap<T, BTreeSet<RefID>>)
```

Reply:
We will revisit this point when we move towards a unified implementation of the borrow graph.